### ISSUE TITLE
[css-flexbox] Non-interop with shrinking images

### ISSUE THREAD

**Issue by jensimmons** [w3c/csswg-drafts#1322]
[Testcase](https://software.hixie.ch/utilities/js/live-dom-viewer/?<!DOCTYPE%20html>%0A<style>%0A%20div%20{%20display%3A%20flex%3B%20border%3A%20solid%3B%20width%3A%208em%3B%20}%0A%20p%20{%20margin%3A%202px%3B%20background%3A%20silver%3B%20flex%3A%201%3B%20text-align%3A%20center%3B%20}%0A<%2Fstyle>%0A<div>%0A%20<img%20src%3D"https%3A%2F%2Fplaceholdit.imgix.net%2F~text%3Ftxtsize%3D33%26txt%3D350%25C3%2597150%26w%3D350%26h%3D150">%0A%20<p>BBBBBB%0A%20<p>C%20C%20C%0A<%2Fdiv>) that @fantasai created as we were discussing Flexbox.

**Comment by jensimmons** [comment]
We have a bunch of images. They have various intrinsic sizes. And they are wrapped with a div. Like this markup:
`<div>`
`  <img>`
`  <img>`
`  <img>`

Now let's make the div into flex container. Making the images flex children.

Perhaps there are only a few images, and they are set to no-wrap.
Perhaps there are many images, and they are set to wrap. 

Now apply variations of the `flex` property to this situation. 

If the images are set to shrink, they should shrink. 
If the images are set to stretch, they should stretch while maintaining the aspect ratio (adjusting height through a transferred size).

Only Firefox gets this right. Every other browser has problems. 

Here, look for yourself: http://labs.jensimmons.com/2017/02-008-bug.html

This is bad. It means Flexbox is not doing a huge part of what it's supposed to do. It doesn't size content properly if that content has an intrinsic size. 

Can we fix this?

**Comment by rachelandrew** [comment]
There is a bunch of discussion here regarding stretching of images with an intrinsic aspect ratio https://github.com/w3c/csswg-drafts/issues/1112

**Comment by fantasai** [comment]
Adjusting description back to focus on the testcase that was initially posted; there's definitely a bunch of bugs in image sizing, but wanted to discuss specifically [this testcase](data:text/html;charset=utf-8;base64,PCFET0NUWVBFIGh0bWw%2BDQo8c3R5bGU%2BDQogZGl2IHsgZGlzcGxheTogZmxleDsgYm9yZGVyOiBzb2xpZDsgd2lkdGg6IDUwJTsgfQ0KIHAgeyBtYXJnaW46IDJweDsgYmFja2dyb3VuZDogc2lsdmVyOyBmbGV4OiAxOyB0ZXh0LWFsaWduOiBjZW50ZXI7IH0NCjwvc3R5bGU%2BDQo8ZGl2Pg0KIDxpbWcgc3JjPSJodHRwczovL3BsYWNlaG9sZGl0LmltZ2l4Lm5ldC9%2BdGV4dD90eHRzaXplPTMzJnR4dD0zNTAlQzMlOTcxNTAmdz0zNTAmaD0xNTAiPg0KIDxwPkJCQkJCQg0KIDxwPkMgQyBDDQo8L2Rpdj4%3D) which shows some very interesting non-conforming behavior in Firefox.

Basically what's happening is that the flex items shrink down in different stages. First of course all the excess space gets eaten up, and then the items shrink down to their min-content sizes (triggering wrapping at the spaces in item C). Then, however, instead of flooring the items at that point, Firefox starts to shrink the image. It seems to have a two-stage concept of the min-content floor.

I don't know how it's doing this, but it seems rather nice, so might be worth discussing if we want to adopt it somehow.

**Comment by jensimmons** [comment]
This example is now also here: https://codepen.io/jensimmons/pen/aVmogq

**Comment by dholbert** [comment]
(I didn't notice the `p { ...flex: 1}` at first -- the behavior makes more sense in light of that.)

So, there's really just one main Firefox quirk here:  we're treating the `img` as having `min-width:0` (i.e. we're resolving its `auto` to `0`).  This is because our implementation is from an older version of the spec text, which explicitly said to resolve min-width:auto to 0 for elements with an intrinsic aspect ratio.  [Bug 1136312](https://bugzilla.mozilla.org/show_bug.cgi?id=1136312) is filed on updating that.  This Firefox behavior will change once we fix that bug.

Here's a version of the codepen with explicit "min-width:0" to make other browsers match that behavior:
https://codepen.io/anon/pen/BmLaZK

With that tweak, Firefox and Edge agree on the rendering, I think. (Chrome doesn't quite agree -- it doesn't honor the aspect ratio at all as the image shrinks, whereas Firefox/Edge honor the aspect ratio up until the point where the "C C C" is the tallest thing and forces the image to that height via `align-self:stretch`)

**Comment by cbiesinger** [comment]
Blink has some bugs with aspect ratios and images in flexbox, that could be part of that. I filed https://crbug.com/781972 today for another issue but there's also existing bugs, and they may all be the same underlying issue.

**Comment by fantasai** [comment]
OK, sounds like we're closing this issue as “implementations just need to fix their bugs”. Please re-open if that's not the case. :)
